app/com/arpnetworking/rollups/MetricsDiscovery.java

```diff
@@ -0,0 +1,117 @@
+import com.arpnetworking.play.configuration.ConfigurationHelper;
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+import com.typesafe.config.Config;
+import scala.concurrent.duration.FiniteDuration;
+
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+
+/**
+ * Actor for discovering the list of metrics available to be rolled up on a periodic basis.
+ *
+ * This actor maintains an internal set of metric names and acts as a source for downstream
+ * actors that perform the actual rollups.  This is intended to be used as a singleton in the
+ * cluster.
+ *
+ * @author Gilligan Markham (gmarkham at dropbox dot com)
+ */
+public class MetricsDiscovery extends AbstractActorWithTimers {
+
+    @Override
+    public Receive createReceive() {
+        return receiveBuilder()
+                .matchEquals(FETCH_MSG,
+                        work -> fetchMetricsForRollup())
+                .match(KairosMetricNamesQueryResponse.class,
+                        this::updateMetricsSet)
+                .match(Status.Failure.class,
+                        failure -> LOGGER.warn("Failed to get metrics from Kairo", failure.cause()))
+                .match(MetricFetch.class, work -> {
+                    final Optional<String> metricName = getNextMetric();
+                    if (metricName.isPresent()) {
+                        getSender().tell(metricName.get(), getSelf());
+                    } else {
+                        getSender().tell(NoMoreMetrics.getInstance(), getSelf());
+                    }
+                })
+                .build();
+    }
+
+    /**
+     * Metrics discovery constructor.
+     *
+     * @param configuration play configuration object
+     * @param kairosDbClient client to use for fetching metric names
+     */
+    @Inject
+    public MetricsDiscovery(final Config configuration, final KairosDbClient kairosDbClient) {
+        final FiniteDuration fetchInterval = ConfigurationHelper.getFiniteDuration(configuration, "rollup.fetch.interval");
+        _kairosDbClient = kairosDbClient;
+        _metricsSet = new LinkedHashSet<>();
+        getSelf().tell(FETCH_MSG, ActorRef.noSender());
+        getTimers().startPeriodicTimer(HOURLY_TIMER, FETCH_MSG, fetchInterval);
+    }
+
+    private void fetchMetricsForRollup() {
+        PatternsCS.pipe(_kairosDbClient.queryMetricNames(), getContext().dispatcher())
+                .to(getSelf());
+    }
+
+    private void updateMetricsSet(final KairosMetricNamesQueryResponse response) {
+        response.getResults().stream()
+                .filter(ROLLUP_METRIC_RE.asPredicate().negate())
+                .forEach(_metricsSet::add);
+        _setIterator = null;
+    }
+
+    private Optional<String> getNextMetric() {
+        final String next;
+        if (_setIterator == null) {
```

Comment:
Should we protect calls to this if `_metricsSet` is still `null` that they return absent?

Reply:
Alternatively, if you set `_setIterator` in `updateMetricsSet` and clear it when there are none left then a `null` value for `_setIterator` means there is no data.

Reply:
Switched to always have _setIterator defined so that it should never be null.